ui/tutorial-renderer/src/main/java/org/cru/godtools/tutorial/activity/OptInOnBoardingActivity.kt

```diff
@@ -0,0 +1,57 @@
+package org.cru.godtools.tutorial.activity
+
+import android.os.Bundle
+import androidx.appcompat.app.AppCompatActivity
+import androidx.viewpager.widget.ViewPager
+import me.relex.circleindicator.CircleIndicator
+import org.cru.godtools.tutorial.R
+import org.cru.godtools.tutorial.adapter.OnBoardingPagerAdapter
+import org.cru.godtools.tutorial.util.OnBoardingCallbacks
+
+class OptInOnBoardingActivity : AppCompatActivity(),
+    OnBoardingCallbacks {
+    private lateinit var viewPager: ViewPager
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_optin_onboarding)
+        viewPager = findViewById(R.id.onboarding_viewpager)
+        viewPager.adapter = OnBoardingPagerAdapter(this).also {
+            it.onBoardingPagesLayout = listOf(
+                R.layout.optin_onboarding_explore_slide,
+                R.layout.optin_onboarding_prepare_slide,
+                R.layout.optin_onboarding_try_slide,
+                R.layout.optin_onboarding_menu_slide
+            )
+            it.notifyDataSetChanged()
```

Comment:
any `notifyDataSetChanged()` calls should be internal to the adapter, not from an external source

Reply:
triggering it from an external source has 1 of 2 issues. it's either unnecessary, or prone to be forgotten. the adapter should be the component that is in charge of knowing when it's data was updated.